### ISSUE TITLE
Question about LayerNorm in Wav2vec

### ISSUE THREAD

**Issue by bofenghuang** [speechbrain/speechbrain#1569]
Hi @TParcollet ,

I have a question about the layer normalization used in wav2vec for both HF and Fairseq implementations. Why using it on all `(batch_size, sequence_length, hidden_size)` dimensions of the wav2vec `last_hidden_state` output instead of just on the `hidden_size` dimension, as showed in the examples of pytorch doc [here](https://pytorch.org/docs/stable/generated/torch.nn.LayerNorm.html) (maybe better call it `InstanceNorm`).

Thanks in advance !

https://github.com/speechbrain/speechbrain/blob/424e7921531b0ea6523557ef0fd6ca249936bd26/speechbrain/lobes/models/huggingface_wav2vec.py#L280-L285

https://github.com/speechbrain/speechbrain/blob/424e7921531b0ea6523557ef0fd6ca249936bd26/speechbrain/lobes/models/fairseq_wav2vec.py#L172-L176

**Comment by TParcollet** [comment]
Hi !

I think you are right, I did a mistake like this once if I remember correctly. Do you already know if this has an impact? I don't feel like retraining a model with the change as it is extremely expensive. Could you also provide the correct implementation example ? (With a PR it would be awesome!!!) 

Thanks again !

**Comment by bofenghuang** [comment]
Hi,

Thanks for your response. I didn't test it yet. Let me first run an ASR recipe on it before making the PR :)

**Comment by bofenghuang** [comment]
Hello again,

I came back with several results on LibriSpeech with this [ctc recipe](https://github.com/speechbrain/speechbrain/blob/424e7921531b0ea6523557ef0fd6ca249936bd26/recipes/LibriSpeech/ASR/CTC/hparams/train_with_wav2vec.yaml). The new `layer_norm` of wav2vec output has been done only on the last `hidden_size` dimension.

I also found that the normalization of wav is done across the wavs in the current code.

https://github.com/speechbrain/speechbrain/blob/424e7921531b0ea6523557ef0fd6ca249936bd26/speechbrain/lobes/models/huggingface_wav2vec.py#L277-L278

But in HF this has been done on each wav. So I also ran some tests with this setup.

https://github.com/huggingface/transformers/blob/a26114777ee1c2802e91bd9cb26a3b39974d52ba/src/transformers/models/wav2vec2/feature_extraction_wav2vec2.py#L81-L100

Here are the results with `batch_size=6` and  `batch_size=8`.

```
# -- bs6
# current version
Test-clean: %WER 2.05 [ 1079 / 52576, 158 ins, 88 del, 833 sub ]
Test-other: %WER 4.18 [ 2187 / 52343, 307 ins, 131 del, 1749 sub ]
# w/ new layernorm
Test-clean: %WER 1.91 [ 1005 / 52576, 114 ins, 63 del, 828 sub ]
Test-other: %WER 4.00 [ 2092 / 52343, 258 ins, 100 del, 1734 sub ]
# w/ new layernorm, new normalization
Test-clean: %WER 1.87 [ 984 / 52576, 97 ins, 74 del, 813 sub ]
Test-other: %WER 3.92 [ 2053 / 52343, 217 ins, 114 del, 1722 sub ]

# -- bs8
# current version
Test-clean: %WER 1.92 [ 1012 / 52576, 104 ins, 75 del, 833 sub ]
Test-other: %WER 3.99 [ 2088 / 52343, 244 ins, 114 del, 1730 sub ]
# w/ new layernorm
Test-clean: %WER 1.97 [ 1037 / 52576, 108 ins, 78 del, 851 sub ]
Test-other: %WER 4.10 [ 2148 / 52343, 279 ins, 121 del, 1748 sub ]
# w/ new layernorm, new normalization
Test-clean: %WER 2.04 [ 1071 / 52576, 90 ins, 119 del, 862 sub ]
Test-other: %WER 4.00 [ 2092 / 52343, 202 ins, 140 del, 1750 sub ]
```

The impact seems to be completely different on two batch sizes. What's your idea about that @TParcollet ?

**Comment by TParcollet** [comment]
Hehehe no idea! The thing being that currently, the batch size impacts decoding (this will change with the next major release). But I think that we should remain closer to the original implementation ...

**Comment by bofenghuang** [comment]
Should we make a PR about new layer normalization and wav normalization? The model trained with these changes got 1.87% WER, slightly better than the model in [README](https://github.com/speechbrain/speechbrain/blob/424e7921531b0ea6523557ef0fd6ca249936bd26/recipes/LibriSpeech/ASR/CTC/README.md), using only one epoch.

Looking forward to the new release :)

**Comment by TParcollet** [comment]
Yes please !

**Comment by TParcollet** [comment]
(we could use the ckpt of the model as well to update HF and GDRIVE).

**Comment by TParcollet** [comment]
1. It can be done in the dataloader. But I already noted that we should add our padded-compatible layernorm on these recipes. Honestly, I wouldn't expect this to impact to much the results ... It's adding regularization.
